src/tonic/api/points_api.rs

```diff
@@ -96,6 +256,137 @@ impl TryFrom<PointStruct> for collection::operations::point_ops::PointStruct {
+                }),
+            ),
+            (None, Some(FilterSelector { filter: Some(f) })) => Ok(
+                collection::operations::point_ops::PointsSelector::FilterSelector(
+                    collection::operations::point_ops::FilterSelector {
+                        filter: f.try_into()?,
+                    },
+                ),
+            ),
+            _ => Err(Status::failed_precondition("Malformed PointsSelector type")),
+        }
+    }
+}
+
+fn none_if_empty<A>(vec: Vec<A>) -> Option<Vec<A>> {
+    if vec.is_empty() {
+        None
+    } else {
+        Some(vec)
+    }
+}
+
+impl TryFrom<Filter> for segment::types::Filter {
+    type Error = Status;
+
+    fn try_from(value: Filter) -> Result<Self, Self::Error> {
+        let should: Vec<_> = value
+            .should
+            .into_iter()
+            .map(|c| c.try_into())
+            .collect::<Result<_, _>>()?;
+        let must: Vec<_> = value
+            .must
+            .into_iter()
+            .map(|c| c.try_into())
+            .collect::<Result<_, _>>()?;
+        let must_not: Vec<_> = value
+            .must_not
+            .into_iter()
+            .map(|c| c.try_into())
+            .collect::<Result<_, _>>()?;
+        Ok(Self {
+            should: none_if_empty(should),
+            must: none_if_empty(must),
+            must_not: none_if_empty(must_not),
+        })
+    }
+}
+
+impl TryFrom<Condition> for segment::types::Condition {
+    type Error = Status;
+
+    fn try_from(value: Condition) -> Result<Self, Self::Error> {
+        match (value.field, value.has_id, value.filter) {
+            (Some(field), None, None) => Ok(segment::types::Condition::Field(field.try_into()?)),
+            (None, Some(has_id), None) => Ok(segment::types::Condition::HasId(has_id.try_into()?)),
+            (None, None, Some(filter)) => Ok(segment::types::Condition::Filter(filter.try_into()?)),
+            _ => Err(Status::failed_precondition("Malformed Condition type")),
+        }
+    }
+}
+
+impl TryFrom<HasIdCondition> for segment::types::HasIdCondition {
+    type Error = Status;
+
+    fn try_from(value: HasIdCondition) -> Result<Self, Self::Error> {
+        let vec: Vec<PointIdType> = value.has_id.into_iter().map(|p| p.into()).collect();
+        let mut set = HashSet::new();
+        for v in vec {
+            set.insert(v);
+        }
+        Ok(Self { has_id: set })
+    }
+}
+
+impl TryFrom<FieldCondition> for segment::types::FieldCondition {
+    type Error = Status;
+
+    fn try_from(value: FieldCondition) -> Result<Self, Self::Error> {
+        match value {
+            FieldCondition {
+                key: Some(k),
+                r#match,
+                range,
+                geo_bounding_box,
+                geo_radius,
+            } => {
+                let geo_bounding_box = geo_bounding_box.map(|g| g.try_into().unwrap()); //TODO why can't use `?`
```

Comment:
Any idea why can't I use `?` within `Option.map` here?

Reply:
Maybe something like 

```
geo_bounding_box.map_or_else(Ok(None), |g| g.try_into())?;
```

?

Reply:
Thanks for the suggestion.

> geo_bounding_box.map_or_else(Ok(None), |g| g.try_into())?;

Unfortunately it does not work but this version works for instance:

```rust
let geo_bounding_box = geo_bounding_box.map_or_else(|| Ok(None), |g| g.try_into().map(Some))?;
```

It is folding the `Option` into a `Result` and re-wrapping as an `Option`.

Reply:
https://github.com/qdrant/qdrant/pull/264/commits/c198d0c4710e2cb82c9e6715e0919e3ff5535d3b